File: 1260.cpp
#include <iostream>
#include <queue>
using namespace std;

#define MAX 1001

int N, M, V; // vertex 개수, edge 개수, start vertex
bool vertex[MAX][MAX] = { false, };
bool visited[MAX] = { false, };
queue<int> q;

void DFS(int from) {
	cout << from << " ";
	visited[from] = true;
	for (int i = 1; i <= N; i++) {
		if (vertex[from][i] && !visited[i]) {
			visited[i] = true;
			DFS(i);
		}
	}
}

void BFS(int from) {
	visited[from] = true;
	q.push(from);
	while (!q.empty()) {
		int num = q.front();
		cout << num << " ";
		q.pop();
		for (int i = 1; i <= N; i++) {
			if (vertex[num][i] && !visited[i]) {
				q.push(i);
				visited[i] = true;
			}
		}
	}
}

int main() {
	cin >> N >> M >> V;
	for (int i = 0; i < M; i++) {
		int from, to;
		cin >> from >> to;
		vertex[from][to] = vertex[to][from] = true; // 양방향
	}

	DFS(V);
	cout << endl;
	for (int i = 1; i <= N; i++)
		visited[i] = false; // init
	BFS(V);
}

File: BOJ/1260.py
n,m,v = map(int, input().split())

vertex = [[False] * (n+1) for _ in range(n+1)] # vertex = [[False] * (n+1)] * (n+1)는 리스트 안의 요소 리스트들이 모두 같은 참조값을 가짐
visited = [False] * (n+1)

for i in range(m):
    start, to = map(int, input().split())
    vertex[start][to] = vertex[to][start] = True

def DFS(v):
    visited[v] = True
    print(v, end=' ')

    for i in range(1, n+1):
        if vertex[v][i] and visited[i] == False:
            DFS(i)

DFS(v)
print()

visited = [False] * (n+1)
stack = []
def BFS(v):
    stack.append(v)
    visited[v] = True

    while len(stack) != 0:
        vert = stack.pop(0)
        print(vert, end=' ')
        for i in range(1, n+1):
            if vertex[vert][i] and visited[i] == False:
                stack.append(i)
                visited[i] = True

BFS(v)
    
